### ISSUE TITLE
Transfer learning options

### ISSUE THREAD

**Issue by cri5Castro** [facebookresearch/wav2letter#507]
Hi there I have been going to the doc and try to find a way to apply transfer learning, my question is there is any option to freeze some parts of the AM  models ?, I saw that there is "fork" mode in for training (which I imagine is used for full finetuning) but I can find any flag to indicate the trained model binary

**Comment by vineelpratap** [comment]
Hi,
We don't support it off-the-shelf but it is possible to do fine-tuning in wav2letter++ by making some code changes.  Just add the following function 
```
// nLayersForFinetuning - number of layers (with parameters) starting from the last layer that we need to fine tune
void setTrainForFinetuning(std::shared_ptr<fl::Module> ntwrk, int nLayersForFinetuning) {
  if (nLayersForFinetuning < 0) {
    ntwrk->train();
    return;
  }
  auto seq = std::dynamic_pointer_cast<fl::Sequential>(ntwrk);
  if (!seq) {
    throw std::runtime_error("something went wrong.");
  }
  int processedLastLayers = 0;
  for (int i = seq->modules().size() - 1; i >= 0; --i) {
    auto module = seq->module(i);
    if (module->params().size() > 0) {
      if (processedLastLayers < nLayersForFinetuning) {
        processedLastLayers++;
        module->train();
      } else {
        module->eval();
      }
    }
  }
}
```

and replace `ntwrk->train()` in `Train.cpp` file with `setTrainForFinetuning(ntwrk, nLayersForFinetuning)` and the run the training in fork mode. 

Hope it helps !

**Comment by cri5Castro** [comment]
great! thanks for the help , I imagine I will also need to recompile the code right?, in the other hand is posible to perform finetuning using the fork mode (I mean using a pre-trained model and changing the output dim to the token size in my problem)

**Comment by vineelpratap** [comment]
> I imagine I will also need to recompile the code right?

Yes, make sure you are using the latest code, make the code changes as mentioned above and recompile. 

> changing the output dim to the token size in my problem

Yes, it is possible! But it would need some more code changes for your specific use case. If you can mention the architecture, model being fine-tuned and details on the tokens that you want to change, I can give some code pointers.

**Comment by cri5Castro** [comment]
that would be very nice thanks, I'm using the original [conv_glu (https://github.com/facebookresearch/wav2letter/blob/master/recipes/models/conv_glu/librispeech/network.arch) because of the relatively small number of parameters for my AM and i'm using wordpiece tokenizer with 9996 different tokens

**Comment by vineelpratap** [comment]
For running the fork model, 

- modify https://github.com/facebookresearch/wav2letter/blob/master/Train.cpp to 

<img width="881" alt="Screen Shot 2020-01-31 at 8 36 45 AM" src="https://user-images.githubusercontent.com/5282102/73556749-e32a9c80-4404-11ea-89f5-777e93b90240.png">

- Run the model with 
> Train fork <MODEL_PATH> --tokensdir '' --tokensfile <PATH_TO_NEW_TOKEN_FILE> --lexicon <PATH_TO_NEW_LEXICON_FILE> --archdir '' --arch <PATH_TO_ARCH_FILE> -- other necessary gflags

Note that this is rough way to do this. You might have to adapt depend on your specific use case. 


FWIW, I wouldn't recommend to use the `conv_glu` model for token size `9996` because of two reasons
- It is trained with ASG criterion. Using such high token size will make the training very slow. 
- The total stride of the architecture is 2. For such high token size, we keep total stride of the architecture 8 for making the training faster.

**Comment by cri5Castro** [comment]
thanks for the recomendations ;)

**Comment by cri5Castro** [comment]
so you recommend me to use word in char level if i want to use GLU with ASG right

**Comment by vineelpratap** [comment]
yes.

**Comment by raghuch** [comment]
Hi, We are trying to do transfer learning from librivox SOTA recipe using TDS Seq2Seq (as mentioned here: https://github.com/facebookresearch/wav2letter/tree/master/recipes/models/sota/2019), with 4 warmup epochs and stepsize 60 epochs on our private dataset with seq2seq criterion. (creating the lexicon and tokens and wordpiece in the usual way given by the recipe).

However, even after 7-8 epochs, the train WER still stays high (say, 100 to 105) and doesn't change. What could be causing the issue that the WER doesn't come down even a bit?

Also, are there some recommended config settings for better transfer learning?

**Comment by NatSp** [comment]
> For running the fork model,
> 
> * modify https://github.com/facebookresearch/wav2letter/blob/master/Train.cpp to
> 
> <img alt="Screen Shot 2020-01-31 at 8 36 45 AM" width="881" src="https://user-images.githubusercontent.com/5282102/73556749-e32a9c80-4404-11ea-89f5-777e93b90240.png">
> 
> * Run the model with
> 
> > Train fork <MODEL_PATH> --tokensdir '' --tokensfile <PATH_TO_NEW_TOKEN_FILE> --lexicon <PATH_TO_NEW_LEXICON_FILE> --archdir '' --arch <PATH_TO_ARCH_FILE> -- other necessary gflags
> 
> Note that this is rough way to do this. You might have to adapt depend on your specific use case.
> 
> FWIW, I wouldn't recommend to use the `conv_glu` model for token size `9996` because of two reasons
> 
> * It is trained with ASG criterion. Using such high token size will make the training very slow.
> * The total stride of the architecture is 2. For such high token size, we keep total stride of the architecture 8 for making the training faster.

Hello, I am trying to implement transfer learning as you showed but when I run the code I get the following error:

I0526 13:44:57.753671 10774 Train.cpp:217] Loading architecture file from /home/john/Downloads/asr_models/am.arch
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Invalid input size:203):
In function af::dim4 getOutDims(const af::dim4&, const af::dim4&, bool)
In file src/backend/common/ArrayInfo.cpp:134
Invalid dimension for argument 1
Expected: ldims == rdims
 0# 0x00007F401A1C7790 in /usr/local/lib/libafcuda.so.3
 1# af_div in /usr/local/lib/libafcuda.so.3
 2# af::operator/(af::array const&, af::array const&) in /usr/local/lib/libafcuda.so.3
 3# 0x000055867573DA8E in ./Train
 4# 0x000055867578ED73 in ./Train
 5# 0x0000558675498469 in ./Train
 6# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
 7# 0x00005586754D70CE in ./Train

In function af::array af::operator/(const af::array&, const af::array&)
In file src/api/cpp/array.cpp:921
*** Aborted at 1590489898 (unix time) try "date -d @1590489898" if you are using GNU date ***
PC: @     0x7f3ff7f893eb gsignal
*** SIGABRT (@0x3e800002a16) received by PID 10774 (TID 0x7f3fd23c7d80) from PID 10774; stack trace: ***
    @     0x7f3ff8892641 (unknown)
    @     0x7f40216f6540 (unknown)
    @     0x7f3ff7f893eb gsignal
    @     0x7f3ff7f68899 abort
    @     0x7f3ff833d5f6 (unknown)
    @     0x7f3ff83499ec (unknown)
    @     0x7f3ff8349a47 std::terminate()
    @     0x7f3ff8349ca9 __cxa_throw
    @     0x7f401a11e413 af::operator/()
    @     0x55867573da8e fl::operator/()
    @     0x55867578ed73 fl::WeightNorm::computeWeight()
    @     0x558675498469 main
    @     0x7f3ff7f6a1e3 __libc_start_main
    @     0x5586754d70ce _start
Aborted (core dumped)

I am using the lexicon_free arch file of wsj.
